crates/ruff/src/autofix/mod.rs

```diff
@@ -59,47 +59,49 @@ fn apply_fixes<'a>(
         })
         .sorted_by(|(rule1, fix1), (rule2, fix2)| cmp_fix(*rule1, *rule2, fix1, fix2))
     {
-        // If we already applied an identical fix as part of another correction, skip
-        // any re-application.
-        if fix.edits().iter().all(|edit| applied.contains(edit)) {
-            *fixed.entry(rule).or_default() += 1;
-            continue;
-        }
+        // Remove any edits that were applied as part of a previous fix.
+        let edits = fix.edits();
+
+        let mut first = true;
+        'inner: for edit in edits.to_sorted() {
+            // Skip any edits that were already applied.
+            if applied.contains(&edit) {
+                continue 'inner;
+            }
 
-        // Best-effort approach: if this fix overlaps with a fix we've already applied,
-        // skip it.
-        if last_pos.map_or(false, |last_pos| {
-            fix.min_start()
-                .map_or(false, |fix_location| last_pos >= fix_location)
-        }) {
-            continue;
-        }
+            // Lazily enforce any isolation and positional requirements (e.g., avoid applying
+            // overlapping fixes, but avoid applying this requirement if all fixes in the edit were
+            // already applied).
+            if first {
+                // If this fix requires isolation, and we've already applied another fix in the
+                // same isolation group, skip it.
+                if let IsolationLevel::Group(id) = fix.isolation() {
+                    if !isolated.insert(id) {
+                        continue 'outer;
+                    }
+                }
+
+                // Best-effort approach: if this fix overlaps with a fix we've already applied,
+                // skip it.
+                if last_pos.map_or(false, |last_pos| last_pos >= edit.start()) {
+                    continue 'outer;
+                }
 
-        // If this fix requires isolation, and we've already applied another fix in the
-        // same isolation group, skip it.
-        if let IsolationLevel::Group(id) = fix.isolation() {
-            if !isolated.insert(id) {
-                continue;
+                first = false;
```

Comment:
This is kind of confusing (definitely easier to read locally than as a diff), but in short: we want to apply these constraints at the fix level, but if all of the fix's edits were already applied, we don't need to enforce them at all. So we do so lazily, only checking if the location is acceptable when we reach the first edit-to-apply.